src/core/Akka.Streams/Implementation/Fusing/StreamOfStreams.cs

```diff
@@ -1,576 +1,809 @@
+            Shape = new FlowShape<T, Tuple<IEnumerable<T>, Source<T, Unit>>>(_in, _out);
         }
 
-        private sealed class TailSource : GraphStage<SourceShape<T>>
+        protected override Attributes InitialAttributes { get; } = DefaultAttributes.PrefixAndTail;
+
+        public override FlowShape<T, Tuple<IEnumerable<T>, Source<T, Unit>>> Shape { get; }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) => new PrefixAndTailLogic(this);
+
+        public override string ToString() => $"PrefixAndTail({_count})";
+    }
+
+    /// <summary>
+    /// INTERNAL API
+    /// </summary>
+    internal static class Split
+    {
+        internal enum SplitDecision
         {
-            private sealed class TailSourceLogic : GraphStageLogic, ITail
+            SplitBefore,
+            SplitAfter
+        }
+
+        public static IGraph<FlowShape<T, Source<T, Unit>>, Unit> When<T>(Func<T, bool> p, SubstreamCancelStrategy substreamCancelStrategy) where T : class => new Split<T>(SplitDecision.SplitBefore, p, substreamCancelStrategy);
+
+
+        public static IGraph<FlowShape<T, Source<T, Unit>>, Unit> After<T>(Func<T, bool> p, SubstreamCancelStrategy substreamCancelStrategy) where T : class => new Split<T>(SplitDecision.SplitAfter, p, substreamCancelStrategy);
+    }
+
+    /// <summary>
+    /// INTERNAL API
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class Split<T> : GraphStage<FlowShape<T, Source<T, Unit>>> where T : class
+    {
+        #region internal classes
+
+        private sealed class SplitLogic : TimerGraphStageLogic
+        {
+            #region internal classes 
+
+            private sealed class SubstreamHandler : InAndOutHandler
             {
-                private readonly TailSource _stage;
-                private readonly Action<T> _onParentPush;
-                private readonly Action<Unit> _onParentFinish;
-                private readonly Action<Exception> _onParentFailure;
+                private readonly SplitLogic _logic;
+                private bool _willCompleteAfterInitialElement;
 
-                public TailSourceLogic(Shape shape, TailSource stage) : base(shape)
+                public SubstreamHandler(SplitLogic logic)
                 {
-                    _stage = stage;
-                    _onParentPush = GetAsyncCallback<T>(element => Push(stage.Out, element));
-                    _onParentFinish = GetAsyncCallback<Unit>(_ => CompleteStage());
-                    _onParentFailure = GetAsyncCallback<Exception>(FailStage);
-
-                    SetHandler(stage.Out,
-                        onPull: stage.PullParent,
-                        onDownstreamFinish: stage.CancelParent);
+                    _logic = logic;
                 }
 
-                public override void PreStart()
+                public bool HasInitialElement => FirstElem != null;
+
+                public T FirstElem { private get; set; }
```

Comment:
Please, no abbreviations ;) (I know, scala has them, but we should follow C# style rules)